### PR TITLE
Custom field analysis command

### PR DESCRIPTION
**Pull request by shinde-rahul**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bugfix? | No
| New feature? |  Yes
| Automated tests included? | Yes
| Related user documentation PR URL |
| Related developer documentation PR URL |
| Issues addressed (#s or URLs) |
| BC breaks? | No
| Deprecations? | No

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
The following is the user story this PR addresses.  
_As an_ **engineer**
_I want to_ **Analyse the actual usage of columns (custom fields) in the leads table**
_So that I can_ **set the appropriate size for VARCHAR columns**.

Whenever the user creates any of the fields, the column gets added to the leads table. The default length for a few fields is 64 characters, and it's 255 characters for a few. The following table summarises it,

| Field Type | DB type | Length
|------------| ------- | -----
| Timezone   | VARCHAR | 255
| Locale     | VARCHAR | 255
| Country    | VARCHAR | 255
| Email      | VARCHAR | 64
| Lookup     | VARCHAR | 255
| Select     | VARCHAR | 64
| Region     | VARCHAR | 255
| Phone      | VARCHAR | 255
| Text       | VARCHAR | 64

This command help analyzes all custom fields (type listed in the above table) and suggests the new length.
The command will generate comma-separated values that will help one to store it as a CSV  file. Otherwise, use `-t` to see the output in tabular format.
The headings in the output,
- Custom Field Name: Field label
- Custom Field Alias: Alias of the field
- Current Size: The character length of the fields (varchar length)
- Currently Utilised max size: Max character length of actual field/column record.
- Suggested max size: Minimum of _Currently Utilised max size" * 2_ or _Current Size_ or 191

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create custom fields
3. Fill in data for custom fields
4. ssh to the webhead,
5. run `bin/console mautic:fields:analyse -t` to see the tabular output
6. Or run `bin/console mautic:fields:analyse > /path/to/csv/file.csv` to save output as CSV file.

[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. Functional test for the command
2. functional test for FieldModel and LeadModel

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10775"></a>

### CODE REVIEW

**Comment by kuzmany** [comment]
Looks like nice PR. 
Return error `Call to an undefined method Mautic\LeadBundle\Entity\LeadField::getCharLengthLimit().`

File: app/bundles/LeadBundle/Translations/en_US/messages.ini [hunk 1]
```diff
@@ -858,6 +858,7 @@ mautic.lead.field.notification.cannot_be_created_message="Custom field '%label%'
 mautic.lead.field.notification.cannot_be_created_header="Custom field cannot be created"
 mautic.lead.field.notification.custom_field_limit_hit_message="Custom field '%label%' cannot be created - you have reached the limit of custom fields allowed by your database."
 mautic.lead.field.notification.custom_field_limit_hit_header="Custom field cannot be created"
+<<<<<<< HEAD
```
Comment (shinde-rahul) [inline app/bundles/LeadBundle/Translations/en_US/messages.ini]: Need to remove this.

File: app/bundles/LeadBundle/Translations/en_US/messages.ini [hunk 2]
```diff
@@ -867,3 +868,14 @@ mautic.lead.stats.click_through_open_rate="Click-to-open rate:"
...
+mautic.lead.segments.usages="Segment usages"
+mautic.lead.campaign.share="Campaign share"
+mautic.lead.share="Share"
+mautic.lead.field.analyse.header.name="Custom Field Name"
+mautic.lead.field.analyse.header.alias="Custom Field Alias"
+mautic.lead.field.analyse.header.length="Current Size"
+mautic.lead.field.analyse.header.max_length="Currently Utilised max size"
+mautic.lead.field.analyse.header.suggested_length="Suggested max size"
+mautic.lead.field.analyse.header.indexed="Is Index Present"
+>>>>>>> c60b4bbd81 (Added a command to analyse the custom fields.)
```
Comment (shinde-rahul) [inline app/bundles/LeadBundle/Translations/en_US/messages.ini]: Needs update.

**Review by escopecz — APPROVED** [review]
The code looks great. Lots of tests added. And the command works also nicely:
```
bin/console mautic:fields:analyse -t
+-------------------+--------------------+--------------+-----------------------------+--------------------+------------------+
| Custom Field Name | Custom Field Alias | Current Size | Currently Utilised max size | Suggested max size | Is Index Present |
+-------------------+--------------------+--------------+-----------------------------+--------------------+------------------+
| Tiny field        | tiny_field         | 10           | 0                           | 10                 | Yes              |
+-------------------+--------------------+--------------+-----------------------------+--------------------+------------------+
```

**Review by kuzmany — APPROVED** [review]
It works

**Comment by RCheesley** [comment]
@shinde-rahul any possibility you can add the docs on https://github.com/mautic/user-documentation/blob/5.x/docs/contacts/custom_fields.rst to include this command? cc @mautic/education-team for visibility, this is required for the 5.1 release.

**Comment by RCheesley** [comment]
Just starting to work on this one!

**Comment by RCheesley** [comment]
https://github.com/mautic/user-documentation/pull/289 docs here.

**Pull request merged**